### ISSUE TITLE
Regression with type parameter shadowing

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#32639]
The following code compiles on stable/beta but fails to compile on nightly

```rust
pub enum T {       
    NonSpecific,   
}                  
                   
fn foo<T>() {      
    T::NonSpecific;
}                  
```

This is minimized from a regression in the [quire crate](https://github.com/tailhook/rust-quire). This looks pretty intentional, however, so I just wanted to confirm! Was this intentional, and are we comfortable letting this slide through?

cc @jseyfried

**Comment by jseyfried** [comment]
Yeah, this was intentional -- it was changed in #32141 to fix #32120.
Could there be more breakage (i.e. was this not detected in a crater run)?

**Comment by alexcrichton** [comment]
This may be all the breakage we're looking at, that was just a crater run between stable and nightly last night, and it was the only error its kind that showed up.

**Comment by jseyfried** [comment]
Ok, I'll write a PR to fix the breakage in that crate. I could also do a warning cycle pretty easily but it looks like that won't be needed.

**Comment by pnkfelix** [comment]
(Yes this is expected to break due to aforementioned fix.)
